### PR TITLE
[FLINK-21393] [formats] Implement ParquetAvroInputFormat

### PR DESCRIPTION
**Pull request by echauchot**
## What is the purpose of the change

*(For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)*

Implement ParquetAvroInputFormat

## Verifying this change

This change added tests and can be verified as follows:
new ParquetAvroInputFormatTest test both simple record and projected fields with simple record but sets parquet.avro.write-old-list-structure to true because AvroRowSerializationSchema does not work with parquet.avro.write-old-list-structure set to false 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes, flink-parquet now depends on flink-avro
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes adds new API  (new input format)
  - The serializers: not really, just make AvroRowSerializationSchema#convertRowToAvroRecord public
  - The runtime per-record code paths (performance sensitive): yes as ParquetInputFormat#convert is implemented
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? javadoc with code example

### CODE REVIEW

**Comment by echauchot** [comment]
R: @HuangZhenQiu 
CC: Dominik Wosiński (don't know your github account)

**Comment by echauchot** [comment]
I forgot to mention that I submitted ParquetAvroInputFormat to master (1.13) but it is made to work for 1.12.x (last release) and I'm using it with Flink 1.12.x.
Maybe it could be a good candidate to be included in an upcoming 1.12.3 release, WDYT ?

**Review by HuangZhenQiu — APPROVED** [review]

**Comment by HuangZhenQiu** [comment]
@JingsongLi Please have another round of review.

**Comment by echauchot** [comment]
Forgot to add deep copy in my field projection impl (in case avro schema is multi-level)

**Comment by Wosin** [comment]
Hey :) Thanks for CC. Looks good! 
One thing to consider, is if there is anything stopping us from going even futher and adding possibility to deserialize `SpecificRecord` instances too ?

**Comment by echauchot** [comment]
> Hey :) Thanks for CC. Looks good!
> One thing to consider, is there is anything stopping us from going even futher and adding possibility to deserialize `SpecificRecord` instances too ?

Nothing I guess, It is just that this PR relies on a already existing _AvroRowSerializationSchema_ that produces GenericRecords from internal flink Row.

**Comment by echauchot** [comment]
@JingsongLi @HuangZhenQiu I just added field deep copy (in case the source schema is multi-level) and fixed serialization issues  that I found testing on flink 1.11

**Comment by DomWos** [comment]
> > Hey :) Thanks for CC. Looks good!
> > One thing to consider, is there is anything stopping us from going even futher and adding possibility to deserialize `SpecificRecord` instances too ?
> 
> Nothing I guess, It is just that this PR relies on a already existing _AvroRowSerializationSchema_ that produces GenericRecords from internal flink Row.

IMHO, we should try to incorporate that (if not too hard obviously), since having at least some type safety would be very beneficial.

**Comment by echauchot** [comment]
@JingsongLi WDYT about this PR?

**Comment by echauchot** [comment]
> > > Hey :) Thanks for CC. Looks good!
> > > One thing to consider, is there is anything stopping us from going even futher and adding possibility to deserialize `SpecificRecord` instances too ?
> > 
> > 
> > Nothing I guess, It is just that this PR relies on a already existing _AvroRowSerializationSchema_ that produces GenericRecords from internal flink Row.
> 
> IMHO, we should try to incorporate that (if not too hard obviously), since having at least some type safety would be very beneficial.

Sure, we can do this in next PR. I'd like that feature to be merged quite quickly so that my project could depend on a snapshot jar (and a released jar in the future) and not on my locally patched jar.

**Comment by Wosin** [comment]
> > > > Hey :) Thanks for CC. Looks good!
> > > > One thing to consider, is there is anything stopping us from going even futher and adding possibility to deserialize `SpecificRecord` instances too ?
> > > 
> > > 
> > > Nothing I guess, It is just that this PR relies on a already existing _AvroRowSerializationSchema_ that produces GenericRecords from internal flink Row.
> > 
> > 
> > IMHO, we should try to incorporate that (if not too hard obviously), since having at least some type safety would be very beneficial.
> 
> Sure, we can do this in next PR. I'd like that feature to be merged quite quickly so that my project could depend on a snapshot jar (and a released jar in the future) and not on my locally patched jar.

Sounds reasonable if that's okay for the guys :)
Cheers.

**Comment by echauchot** [comment]
@JingsongLi friendly ping :)

**Comment by JingsongLi** [comment]
@echauchot Thanks for your contribution, sorry for the late reply.
The PR looks good.
I don't know if I should merge this. We already have new File Source interfaces and implementations. If we add new features to the old interface, it will inevitably bring unnecessary maintenance costs.

**Comment by echauchot** [comment]
@JingsongLi thanks for your comments and sorry for the delay of my answer (I was very busy on other things and then off).
I agree it adds an extra maintenance cost but IMHO it is needed in addition to FileSource because FileSource is not available in flink 1.11.2 which is the last version supported by Amazon EMR cluster and we need to support this in Amazon EMR IMHO.

**Comment by echauchot** [comment]
@JingsongLee, echoing what I sent to the Flink ML

I agree that adding a new feature (ParquetAvroInputFormat) to an old source API is a maintenance burden. But IMHO I think that while the new DataStream batch/streaming convergent API is not 100% functional we still need to maintain older sources and add missing features to them.

Indeed, I realized that DataStream API in batch mode (1) does not support aggregations yet (2) so in such a case a user would stick to the DataSet API. And the new FileSource API with ParquetColumnarRowInputFormat is not available in DataSet API (3).

So, long story short, in some cases a user will have no other choice than using ParquetInputFormat and legacy source.

WDYT ?

[1] https://issues.apache.org/jira/browse/FLINK-19316

[2] https://issues.apache.org/jira/browse/FLINK-22587

[3] https://cwiki.apache.org/confluence/display/FLINK/FLIP-27%3A+Refactor+Source+Interface#FLIP27:RefactorSourceInterface-Compatibility,Deprecation,andMigrationPlan

**Comment by echauchot** [comment]
So can we merge this PR ?

**Comment by echauchot** [comment]
ping @JingsongLi

**Comment by echauchot** [comment]
@JingsongLi can we merge the PR? It has been 2 months since submission

**Comment by echauchot** [comment]
rebased on master

File: flink-formats/flink-parquet/pom.xml [hunk 1]
```diff
@@ -45,6 +45,12 @@ under the License.
 			<scope>provided</scope>
 		</dependency>
 
+		<dependency>
+			<groupId>org.apache.flink</groupId>
```
Comment (AHeise) [inline flink-formats/flink-parquet/pom.xml:49]: I was a bit worried that all users that use parquet now also get the Avro dependencies, but as far as I can see they already have all avro dependencies % flink-avro (128 kb) already. Can you double-check if this is true?

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowSerializationSchema.java [hunk 2]
```diff
@@ -157,7 +157,7 @@ public int hashCode() {
 
     // --------------------------------------------------------------------------------------------
 
-    private GenericRecord convertRowToAvroRecord(Schema schema, Row row) {
+    public GenericRecord convertRowToAvroRecord(Schema schema, Row row) {
```
Comment (AHeise) [inline flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowSerializationSchema.java:160]: I think we should factor that to some utility class instead. It feels unnatural that the `ParquetInputFormat` needs to have an instance of `AvroRowSerializationSchema` and we need to expose this previously private method across packages. However, if we need state for this method, I can be persuaded to keep as is.

File: flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetAvroInputFormatTest.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+
+/** Test cases for reading Parquet files and convert parquet records to Avro GenericRecords. */
+@RunWith(Parameterized.class)
```
Comment (AHeise) [inline flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetAvroInputFormatTest.java:45]: Why is this parameterized? Is this test then even executed on azure?

File: flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetAvroInputFormatTest.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+public class ParquetAvroInputFormatTest extends TestUtil {
```
Comment (AHeise) [inline flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetAvroInputFormatTest.java:46]: Please change to static imports. I know this is how the other tests are doing it, but it's bad practice. If it's indeed a larger change, please leave as-is.

**Review by AHeise — CHANGES_REQUESTED** [review]
Hi @echauchot , thank you very much for your contribution and sorry that your PR got stuck.

I'll take over and guide you to a quick merge.
In general, I think we can keep adding formats as long as we do not fully deprecate DataSet API and it's still a longer way to go as you have noticed. This is especially true, if it's just an extension of an existing format.

I left a couple of comments that I'd like you to resolve as well as the compile issues.

**Comment by echauchot** [comment]
@AHeise ParquetInputFormat base class was removed since I submitted my PR hence the compilation issues, commit ce3631af7313855f675e29b8faa386f6e5a2d43c removed it. This commit mentions "Use the filesystem connector with a Parquet format as a replacement". What is the equivalent then ?

**Comment by AHeise** [comment]
> @AHeise ParquetInputFormat base class was removed since I submitted my PR hence the compilation issues, commit [ce3631a](https://github.com/apache/flink/commit/ce3631af7313855f675e29b8faa386f6e5a2d43c) removed it. This commit mentions "Use the filesystem connector with a Parquet format as a replacement". I guess it refers to https://ci.apache.org/projects/flink/flink-docs-master/docs/connectors/table/filesystem/ which is SQL based. But what if our pipeline pipeline does not use SQL but DataSet API ?

It's a good and tough question. I spoke to @twalthr offline and for now the plan is as follows:
* Table API drops old planner and thus, most of the removed code in that commit is dead code.
* Table API will support everything running on DataStream that used to work when it ran on DataSet.
* 1.13 will be the last release with full DataSet support (BatchTableEnvironment will be dropped)

Since a few features of DataSet are still not supported in DataStream, we expect users to stick to 1.13 for a longer time and probably skip 1.14. So, we'd suggest to merge your 2 PRs to release-1.13 instead of master. Then, all DataSet users would benefit from your contributions while we unblock future developments that would break DataSet as of 1.13.

If it turns out that the community wants to have these features in 1.14 for some reasons (Table API not as far as it should), we can still re-add `ParquetInputFormat` and forward port your PR before feature freeze in 3 months.

Note 1: It might still be possible to have a Flink 1.14 DataSet application using 1.13 formats. In general, it's always possible to copy the old code into your own project.
Note 2: If you are missing combineable aggregations in DataStream, maybe it would be better to move to Table API to begin with. At this point, no-one really knows how much of DataSet will be ported to DataStream. It doesn't really make sense to have 2 APIs with high-level primitives like joins. The main idea is to use Table API by default with a rich user experience and go down to DataStream only when needed (timer, user state, ...).

**Comment by echauchot** [comment]
Hi @AHeise, thanks for your detailed comment. 
Merging my 2 PRs to 1.13 looks fine to me, I think backporting to 1.12 might me considered as well as Amazon EMR is still 1.12 based (latest is 1.12.1) and they are likely to move to latest 1.12 in short term releases of EMR.

For the context,  what I'm doing is a benchmark comparing Flink Dataset/Datstream/SQL and other technologies such as Apache Beam. So I'd like to have a working DataSet/DataStream pipeline in addition to my existing working SQL one (using blink planner based on DataStream and BatchTableEnvironment). And I guess there will be some users that will want to stick to DataSet/DataStream if they have existing pipelines that they do not want to re-code but just migrate to next flink version.

File: flink-formats/flink-parquet/pom.xml (hunk 1 above)
Comment (echauchot) [reply flink-formats/flink-parquet/pom.xml:49]: new transitive deps of flink-parquet: 
_+- org.apache.flink:flink-avro:jar:1.12.1:compile
[INFO] |  \- org.apache.avro:avro:jar:1.10.0:compile
[INFO] |     +- com.fasterxml.jackson.core:jackson-core:jar:2.10.1:compile
[INFO] |     \- com.fasterxml.jackson.core:jackson-databind:jar:2.10.1:compile
[INFO] |        \- com.fasterxml.jackson.core:jackson-annotations:jar:2.10.1:compile_

these avro and jackson deps were already transitive deps of hadoop-common
_
**+- org.apache.hadoop:hadoop-common:jar:2.4.1:provided**
[INFO] |  +- org.apache.hadoop:hadoop-annotations:jar:2.4.1:provided
[INFO] |  +- com.google.guava:guava:jar:11.0.2:compile
[INFO] |  +- commons-cli:commons-cli:jar:1.3.1:provided
[INFO] |  +- org.apache.commons:commons-math3:jar:3.5:provided
[INFO] |  +- xmlenc:xmlenc:jar:0.52:provided
[INFO] |  +- commons-httpclient:commons-httpclient:jar:3.1:provided
[INFO] |  +- commons-codec:commons-codec:jar:1.13:provided
[INFO] |  +- commons-io:commons-io:jar:2.7:provided
[INFO] |  +- commons-net:commons-net:jar:3.1:provided
[INFO] |  +- javax.servlet:servlet-api:jar:2.5:compile
[INFO] |  +- org.mortbay.jetty:jetty:jar:6.1.26:provided
[INFO] |  +- org.mortbay.jetty:jetty-util:jar:6.1.26:provided
[INFO] |  +- com.sun.jersey:jersey-core:jar:1.9:provided
[INFO] |  +- com.sun.jersey:jersey-json:jar:1.9:provided
[INFO] |  |  +- org.codehaus.jettison:jettison:jar:1.1:provided
[INFO] |  |  +- com.sun.xml.bind:jaxb-impl:jar:2.2.3-1:provided
[INFO] |  |  +- org.codehaus.jackson:jackson-jaxrs:jar:1.8.3:provided
[INFO] |  |  \- org.codehaus.jackson:jackson-xc:jar:1.8.3:provided
[INFO] |  +- com.sun.jersey:jersey-server:jar:1.9:provided
[INFO] |  |  \- asm:asm:jar:3.1:provided
[INFO] |  +- tomcat:jasper-compiler:jar:5.5.23:provided
[INFO] |  +- tomcat:jasper-runtime:jar:5.5.23:provided
[INFO] |  +- javax.servlet.jsp:jsp-api:jar:2.1:provided
[INFO] |  +- commons-el:commons-el:jar:1.0:provided
[INFO] |  +- commons-logging:commons-logging:jar:1.1.3:provided
[INFO] |  +- net.java.dev.jets3t:jets3t:jar:0.9.0:provided
[INFO] |  |  +- org.apache.httpcomponents:httpclient:jar:4.5.3:compile
[INFO] |  |  +- org.apache.httpcomponents:httpcore:jar:4.4.6:compile
[INFO] |  |  \- com.jamesmurty.utils:java-xmlbuilder:jar:0.4:provided
[INFO] |  +- commons-lang:commons-lang:jar:2.6:compile
[INFO] |  +- commons-configuration:commons-configuration:jar:1.7:provided
[INFO] |  |  +- commons-digester:commons-digester:jar:1.8.1:provided
[INFO] |  |  \- commons-beanutils:commons-beanutils:jar:1.8.3:provided
[INFO] |  +- org.codehaus.jackson:jackson-core-asl:jar:1.8.8:provided
[INFO] |  +- org.codehaus.jackson:jackson-mapper-asl:jar:1.8.8:provided
**[INFO] |  +- org.apache.avro:avro:jar:1.10.0:compile
[INFO] |  |  +- com.fasterxml.jackson.core:jackson-core:jar:2.10.1:compile
[INFO] |  |  \- com.fasterxml.jackson.core:jackson-databind:jar:2.10.1:compile
[INFO] |  |     \- com.fasterxml.jackson.core:jackson-annotations:jar:2.10.1:compile**
_

Comment (echauchot) [reply flink-formats/flink-parquet/pom.xml:49]: > > @AHeise ParquetInputFormat base class was removed since I submitted my PR hence the compilation issues, commit [ce3631a](https://github.com/apache/flink/commit/ce3631af7313855f675e29b8faa386f6e5a2d43c) removed it. This commit mentions "Use the filesystem connector with a Parquet format as a replacement". I guess it refers to https://ci.apache.org/projects/flink/flink-docs-master/docs/connectors/table/filesystem/ which is SQL based. But what if our pipeline pipeline does not use SQL but DataSet API ?
> 
> It's a good and tough question. I spoke to @twalthr offline and for now the plan is as follows:
> 
> * Table API drops old planner and thus, most of the removed code in that commit is dead code.
> * Table API will support everything running on DataStream that used to work when it ran on DataSet.
> * 1.13 will be the last release with full DataSet support (BatchTableEnvironment will be dropped)
> 
> Since a few features of DataSet are still not supported in DataStream, we expect users to stick to 1.13 for a longer time and probably skip 1.14. So, we'd suggest to merge your 2 PRs to release-1.13 instead of master. Then, all DataSet users would benefit from your contributions while we unblock future developments that would break DataSet as of 1.13.
> 
> If it turns out that the community wants to have these features in 1.14 for some reasons (Table API not as far as it should), we can still re-add `ParquetInputFormat` and forward port your PR before feature freeze in 3 months.
> 
> Note 1: It might still be possible to have a Flink 1.14 DataSet application using 1.13 formats. In general, it's always possible to copy the old code into your own project.
> Note 2: If you are missing combineable aggregations in DataStream, maybe it would be better to move to Table API to begin with. At this point, no-one really knows how much of DataSet will be ported to DataStream. It doesn't really make sense to have 2 APIs with high-level primitives like joins. The main idea is to use Table API by default with a rich user experience and go down to DataStream only when needed (timer, user state, ...).


Hi @AHeise, thanks for your detailed comment.
Merging my 2 PRs to 1.13 looks fine to me, I think backporting to 1.12 might me considered as well as Amazon EMR is still 1.12 based (latest is 1.12.1) and they are likely to move to latest 1.12 in short term releases of EMR.

For the context, what I'm doing is a benchmark comparing Flink Dataset/Datstream/SQL and other technologies such as Apache Beam. So I'd like to have a working DataSet/DataStream pipeline in addition to my existing working SQL one (using blink planner based on DataStream and BatchTableEnvironment). And I guess there will be some users that will want to stick to DataSet/DataStream if they have existing pipelines that they do not want to re-code but just migrate to next flink version.

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowSerializationSchema.java (hunk 2 above)
Comment (echauchot) [reply flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowSerializationSchema.java:160]: great suggestion, I was a bit too quick with the instance ! Thanks ! Refactoring to static bunch of methods (basically convert* methods) works so no state needed. I'll extract them to a util class in flink-avro package (as these methods depend on avro) and refer it in flink-avro/AvroRowSerializationSchema and flink-parquet/ParquetAvroInputFormat. But the module dep will remain.

File: flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetAvroInputFormatTest.java (hunk 4 above)
Comment (echauchot) [reply flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetAvroInputFormatTest.java:46]: not large, just a single test, so ok for static imports. Thanks

Comment (echauchot) [reply flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetAvroInputFormatTest.java:46]: I don't like extension of TestUtils either.

File: flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetAvroInputFormatTest.java (hunk 3 above)
Comment (echauchot) [reply flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetAvroInputFormatTest.java:45]: To be honest I just did the same as the other input format tests. If it is not needed I remove it. If I remember correctly the parametrization is for Avro legacy arrays on/off parameter

**Comment by echauchot** [comment]
> Hi @AHeise, thanks for your detailed comment.
> Merging my 2 PRs to 1.13 looks fine to me, I think backporting to 1.12 might me considered as well as Amazon EMR is still 1.12 based (latest is 1.12.1) and they are likely to move to latest 1.12 in short term releases of EMR.
> 
> For the context, what I'm doing is a benchmark comparing Flink Dataset/Datstream/SQL and other technologies such as Apache Beam. So I'd like to have a working DataSet/DataStream pipeline in addition to my existing working SQL one (using blink planner based on DataStream and BatchTableEnvironment). And I guess there will be some users that will want to stick to DataSet/DataStream if they have existing pipelines that they do not want to re-code but just migrate to next flink version.

Should I re-open a PR against latest not released 1.12.x and latest not released 1.13.x ?

**Comment by AHeise** [comment]
> > Hi @AHeise, thanks for your detailed comment.
> > Merging my 2 PRs to 1.13 looks fine to me, I think backporting to 1.12 might me considered as well as Amazon EMR is still 1.12 based (latest is 1.12.1) and they are likely to move to latest 1.12 in short term releases of EMR.
> > For the context, what I'm doing is a benchmark comparing Flink Dataset/Datstream/SQL and other technologies such as Apache Beam. So I'd like to have a working DataSet/DataStream pipeline in addition to my existing working SQL one (using blink planner based on DataStream and BatchTableEnvironment). And I guess there will be some users that will want to stick to DataSet/DataStream if they have existing pipelines that they do not want to re-code but just migrate to next flink version.
> 
> Should I re-open a PR against latest not released 1.12.x and latest not released 1.13.x ?

I'd change this PR to 1.13 (use the edit button of the PR) and get it done first before thinking about a backport (I will use the time to ask AWS how quickly they plan to adopt 1.13).

File: flink-formats/flink-parquet/pom.xml (hunk 1 above)
Comment (AHeise) [reply flink-formats/flink-parquet/pom.xml:49]: Okay doesn't look too bad. Databind is quite hefty. Maybe we should add this dependency as optional in the same way `parquet-avro` is already optional.

File: flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetAvroInputFormatTest.java (hunk 3 above)
Comment (AHeise) [reply flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetAvroInputFormatTest.java:45]: Please check if this is necessary and remove otherwise.

**Comment by echauchot** [comment]
rebased on release-1.13

File: flink-formats/flink-parquet/pom.xml (hunk 1 above)
Comment (echauchot) [reply flink-formats/flink-parquet/pom.xml]: fair enough

**Comment by AHeise** [comment]
Could you please check the compilation issue? It just looks like you need to rename/remove the test method.

**Comment by echauchot** [comment]
> Could you please check the compilation issue? It just looks like you need to rename/remove the test method.

It is a mix up with a commit of another PR, I'll rebase

File: flink-formats/flink-parquet/pom.xml (hunk 1 above)
Comment (echauchot) [reply flink-formats/flink-parquet/pom.xml]: I put flink-avro optional and excluded jackson-databind

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowSerializationSchema.java (hunk 2 above)
Comment (echauchot) [reply flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowSerializationSchema.java]: I extracted conversion methods to a new util called _AvroConversions_ in flink-avro module

I isolated this change into a different commit. Indeed, IMHO we should keep this commit separated for isolation/revert purposes as it concerns a different module (flink-avro). I found that it was not necessary to open a different ticket just for a refactoring such as this. A separate commit seems enough.

File: flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetAvroInputFormatTest.java (hunk 3 above)
Comment (echauchot) [reply flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetAvroInputFormatTest.java]: removed inheritance and converted to static access

File: flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetAvroInputFormatTest.java (hunk 4 above)
Comment (echauchot) [reply flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetAvroInputFormatTest.java]: removed inheritance and converted to static access

File: flink-formats/flink-parquet/pom.xml (hunk 1 above)
Comment (echauchot) [reply flink-formats/flink-parquet/pom.xml]: My bad  jackson-databind is needed for the other tests in the parquet module

**Comment by echauchot** [comment]
@flinkbot run azure

**Comment by echauchot** [comment]
@AHeise I addressed all your comments, I'm ready for another round of review. PATL

**Comment by echauchot** [comment]
@flinkbot run azure

**Review by AHeise — COMMENTED** [review]
I think we are already there. I left two comments:
- Dependencies can be probably made simpler by just using optional flink-avro.
- Commits showed be squashed except for the refactor commit (I noted that explicitly).

Somewhere we need to add documentation that `flink-avro` is optional dependency (and user needs to add it). I'm thinking maybe to add that in javadoc of `ParquetAvroInputFormat`

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroConversions.java [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.TimeZone;
+
+/** Utility class that takes care of conversions between avro and flink types. */
+public class AvroConversions {
```
[the remark this answers is not in the document]
Comment (echauchot) [reply flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroConversions.java:44]: yes commit isolation is that is what I suggested [here](https://github.com/apache/flink/pull/15156#discussion_r642543053). I see we agree on commit isolation rules :smile: 
Do you want that I change the commit message with "[FLINK-21393][formats] Pulling out AvroConversions out of AvroRowSerializationSchema" ?

[the remark this answers is not in the document]
Comment (AHeise) [reply flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroConversions.java:44]: Yes, sorry for opening another thread and not replying on your comment.
Regarding commit message, it's just important to have a ticket number and component in it. If you have a better message, feel free to keep it.

File: flink-formats/flink-parquet/pom.xml [hunk 6]
```diff
@@ -51,10 +51,24 @@ under the License.
...
+					<groupId>com.fasterxml.jackson.core</groupId>
+					<artifactId>jackson-databind</artifactId>
+				</exclusion>
+			</exclusions>
+		</dependency>
+
+		<dependency>
+			<groupId>com.fasterxml.jackson.core</groupId>
+			<artifactId>jackson-databind</artifactId>
+			<scope>runtime</scope>
```
[the remark this answers is not in the document]
Comment (echauchot) [reply flink-formats/flink-parquet/pom.xml:71]: ah, yes, right, I forgot this dep was shaded. Thanks for pointing out ! 
Cleaning and leaving only the optional dep

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroConversions.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (echauchot) [reply flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroConversions.java:44]: no problem don't be sorry.
rephrased the commit to include [formats] and initial class

**Comment by echauchot** [comment]
> I think we are already there. I left two comments:
> 
> * Dependencies can be probably made simpler by just using optional flink-avro.
> * Commits showed be squashed except for the refactor commit (I noted that explicitly).
> 
> Somewhere we need to add documentation that `flink-avro` is optional dependency (and user needs to add it). I'm thinking maybe to add that in javadoc of `ParquetAvroInputFormat`

I did the change in the javadoc

**Comment by echauchot** [comment]
@AHeise I just squashed everything (including last javadoc and pom changes) to only 2 commits with the correct message format.
Feel free to tell me if you need anything else to be able to merge the PR.

**Review by AHeise — APPROVED** [review]
LGTM. I'll merge when AZP is green.

**Comment by echauchot** [comment]
@AHeise thank you very much for taking over the review and for the quick merge ! I really appreciate.

**Comment by AHeise** [comment]
No thank you for your contribution. I wish it was smoother for you.
